corehq/apps/enterprise/iterators.py

```diff
@@ -1,7 +1,197 @@
+from datetime import datetime, timedelta
+from django.utils.translation import gettext as _
+from corehq.apps.es import filters
+from corehq.apps.es.utils import es_format_datetime
+from corehq.apps.es.forms import FormES
+from corehq.apps.enterprise.exceptions import TooMuchRequestedDataError
+from corehq.apps.app_manager.dbaccessors import get_brief_apps_in_domain
+
+
 def raise_after_max_elements(it, max_elements, exception=None):
     for total_yielded, ele in enumerate(it):
         if total_yielded >= max_elements:
             exception = exception or Exception('Too Many Elements')
             raise exception
 
         yield ele
+
+
+class IterableEnterpriseFormQuery:
+    '''
+    A class representing a query that returns its results as an iterator
+    The intended use case is to support queries that cross pagination boundaries
+    '''
+    def __init__(self, account, start_date, end_date, last_domain, last_time, last_id):
+        MAX_DATE_RANGE_DAYS = 100
+        (self.start_date, self.end_date) = resolve_start_and_end_date(start_date, end_date, MAX_DATE_RANGE_DAYS)
+        self.account = account
+        self.last_domain = last_domain
+        self.last_time = last_time
+        self.last_id = last_id
+
+    def execute(self, limit=None):
+        domains = self.account.get_domains()
+
+        it = multi_domain_form_generator(
+            domains,
+            self.start_date,
+            self.end_date,
+            self.last_domain,
+            self.last_time,
+            self.last_id,
+            limit=limit
+        )
+
+        xform_converter = RawFormConverter()
+        return (xform_converter.convert(form) for form in it)
+
+    @classmethod
+    def get_kwargs_from_map(cls, map):
+        last_domain = map.get('domain', None)
+        last_time = map.get('received_on', None)
+        if last_time:
+            last_time = datetime.fromisoformat(last_time)
+        last_id = map.get('id', None)
+        return {
+            'last_domain': last_domain,
+            'last_time': last_time,
+            'last_id': last_id
+        }
+
+    @classmethod
+    def get_query_params(cls, fetched_object):
+        return {
+            'domain': fetched_object['domain'],
+            'received_on': fetched_object['submitted'],
+            'id': fetched_object['form_id']
+        }
+
+
+def resolve_start_and_end_date(start_date, end_date, maximum_date_range):
+    '''
+    Provide start and end date values if not supplied.
+    '''
+    if not end_date:
+        end_date = datetime.utcnow()
+
+    if not start_date:
+        start_date = end_date - timedelta(days=30)
+
+    if end_date - start_date > timedelta(days=maximum_date_range):
+        raise TooMuchRequestedDataError(
+            _('Date ranges with more than {} days are not supported').format(maximum_date_range)
+        )
+
+    return start_date, end_date
+
+
+class RawFormConverter:
```

Comment:
rename to something like `EnterpriseFormReportConverter` or something better? more specific name re: usecase